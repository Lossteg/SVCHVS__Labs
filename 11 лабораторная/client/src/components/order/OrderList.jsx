import React, { useEffect, useState } from "react";
import { getOrder } from "../../api/orderAPI";
import { Typography, Grid, Button } from "@mui/material";
import Card from "./Card";
import jsPDF from "jspdf";
import autoTable from "jspdf-autotable";
import ExcelJS from "exceljs"; // Импортируем exceljs

function OrderList({ header }) {
  const [orders, setOrders] = useState([]);

  useEffect(() => {
    const fetchOrders = async () => {
      try {
        const data = await getOrder();
        setOrders(data);
      } catch (error) {
        console.log("Error fetching order data:", error);
      }
    };

    fetchOrders();
  }, []);

  const handleDownloadPDFClick = () => {
    console.log("Я пытался");
    const tableColumns = [
      { title: "Service", field: "service" },
      { title: "Properties", field: "properties" },
      { title: "Quantity", field: "quantity", type: "numeric" },
      { title: "Total Price", field: "totalPrice", type: "numeric" },
    ];
    const totalPagesExp = "1";
    const doc = new jsPDF();

    const formatProperties = (row) => {
      if (row.properties) {
        return Object.entries(row.properties)
          .map(([key, value]) => `${key}: ${value}`)
          .join(", ");
      }
      return "";
    };

    doc.page = 1;
    doc.text("Table 1", 15, 10);

    const formattedOrders = orders.map((order) => ({
      ...order,
      properties: formatProperties(order),
    }));

    autoTable(doc, {
      theme: "grid",
      headStyles: { fontSize: 10 },
      bodyStyles: { fontSize: 8, fontStyle: "italic" },
      columns: tableColumns.map((col) => ({
        ...col,
        dataKey: col.field,
      })),
      body: formattedOrders,
    });

    if (typeof doc.putTotalPages === "function") {
      doc.putTotalPages(totalPagesExp);
    }
    doc.save("ReportTable.pdf");
  };

  const handleDownloadExcelClick = () => {
    handleExport().then((blob) => {
      const url = URL.createObjectURL(blob);
      const downloadAnchorNode = document.createElement("a");
      downloadAnchorNode.setAttribute("href", url);
      downloadAnchorNode.setAttribute("download", "report.xlsx");
      downloadAnchorNode.click();
      downloadAnchorNode.remove();
    });
  };

  const handleExport = async () => {
    const formatProperties = (row) => {
      if (row.properties) {
        return Object.entries(row.properties)
          .map(([key, value]) => `${key}: ${value}`)
          .join(", ");
      }
      return "";
    };

    const formattedOrders = orders.map((order) => ({
      ...order,
      properties: formatProperties(order),
    }));

    const workbook = new ExcelJS.Workbook();
    const sheet = workbook.addWorksheet("Table Data");

    const tableColumns = [
      { title: "Service", field: "service" },
      { title: "Properties", field: "properties" },
      { title: "Quantity", field: "quantity", type: "numeric" },
      { title: "Total Price", field: "totalPrice", type: "numeric" },
    ];

    // Добавляем заголовки столбцов в таблицу
    tableColumns.forEach((col, index) => {
      sheet.getCell(1, index + 1).value = col.title;
    });

    // Добавляем данные
    formattedOrders.forEach((order, rowIndex) => {
      sheet.addRow();

      // Здесь вычисляем значение для столбца "Properties"
      const propertiesValue =
        order.properties !== undefined ? order.properties : "";
      sheet.getCell(rowIndex + 2, 2).value = propertiesValue;

      // Заполняем остальные ячейки
      tableColumns.forEach((col, colIndex) => {
        if (col.field !== "properties") {
          const cellValue = order[col.field];
          sheet.getCell(rowIndex + 2, colIndex + 1).value = cellValue;
        }
      });
    });

    // Сохраняем в Blob
    const blob = await workbook.xlsx.writeBuffer();
    return new Blob([blob], {
      type: "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet",
    });
  };

  return (
    <div className="main" align="center">
      <Typography
        variant="h1"
        textAlign="center"
        style={{
          marginBottom: "10px",
          color: "black",
          fontSize: "3.7rem",
          fontWeight: "400",
        }}
      >
        {header}
      </Typography>
      <Button
        onClick={handleDownloadPDFClick}
        color="primary"
        variant="contained"
      >
        Download PDF
      </Button>
      <Button
        onClick={handleDownloadExcelClick}
        color="primary"
        variant="contained"
      >
        Download Excel
      </Button>
      <Grid
        container
        spacing={2}
        style={{ flexDirection: "column", justifyContent: "center" }}
      >
        {Array.isArray(orders) && orders.length > 0 ? (
          orders.map((order, index) => (
            <Grid
              item
              key={index}
              xs="auto"
              sm="auto"
              md="auto"
              style={{ padding: "16px 0px 16px 20px" }}
            >
              <Card order={order} />
            </Grid>
          ))
        ) : (
          <p>You've made no orders yet</p>
        )}
      </Grid>
    </div>
  );
}

export default OrderList;
