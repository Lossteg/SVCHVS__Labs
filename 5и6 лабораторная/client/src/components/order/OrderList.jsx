import React, { useEffect, useState } from 'react';
import { getOrder } from '../../api/orderAPI';
import { Typography, Grid } from '@mui/material';
import Card from "./Card";

function OrderList({ header }) {
  const [orders, setOrders] = useState([]);

  useEffect(() => {
    const fetchOrders = async () => {
      try {
        const data = await getOrder();
        setOrders(data);
      } catch (error) {
        console.log("Error fetching order data:", error);
      }
    };

    fetchOrders();
  }, []);

  return (
    <div className='main' align="center">
      <Typography
        variant="h1"
        textAlign="center"
        style={{
          marginBottom: "10px",
          color: "black",
          fontSize: "3.7rem",
          fontWeight: "400",
        }}
      >
        {header}
      </Typography>
      <Grid container spacing={2} style={{ flexDirection: "column", justifyContent: "center" }}>
        {Array.isArray(orders) && orders.length > 0 ? (
          orders.map((order, index) => (
            <Grid
              item
              key={index}
              xs="auto"
              sm="auto"
              md="auto"
              style={{ padding: "16px 0px 16px 20px" }}
            >
              <Card order={order} />
            </Grid>
          ))
        ) : (
          <p>You've made no orders yet</p>
        )}
      </Grid>
    </div>
  )
}

export default OrderList;

