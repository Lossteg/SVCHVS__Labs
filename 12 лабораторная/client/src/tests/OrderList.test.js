import React from 'react';
import { render } from '@testing-library/react';
import OrderList from '../components/order/OrderList';

jest.mock('../api/orderAPI', () => ({
  getOrder: jest.fn().mockResolvedValue([]), 
}));

describe('OrderList component', () => {
  it('matches snapshot', async () => {
    const { asFragment, findByText } = render(<OrderList header="Test Header" />);
    
    await findByText('You\'ve made no orders yet');

    expect(asFragment()).toMatchSnapshot();
  });
});
