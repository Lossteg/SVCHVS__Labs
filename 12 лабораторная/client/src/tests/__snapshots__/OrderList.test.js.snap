// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OrderList component matches snapshot 1`] = `
<DocumentFragment>
  <div
    align="center"
    class="main"
  >
    <h1
      class="MuiTypography-root MuiTypography-h1 css-sqoje6-MuiTypography-root"
      style="margin-bottom: 10px; color: black; font-size: 3.7rem; font-weight: 400;"
    >
      Test Header
    </h1>
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium css-sghohy-MuiButtonBase-root-MuiButton-root"
      tabindex="0"
      type="button"
    >
      Download PDF
      <span
        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </button>
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium css-sghohy-MuiButtonBase-root-MuiButton-root"
      tabindex="0"
      type="button"
    >
      Download Excel
      <span
        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </button>
    <div
      class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2 css-mhc70k-MuiGrid-root"
      style="flex-direction: column; justify-content: center;"
    >
      <p>
        You've made no orders yet
      </p>
    </div>
  </div>
</DocumentFragment>
`;
